### DISCUSSION TITLE
GraphSAGE sampling operation

### DISCUSSION THREAD

**Discussion by 1234238** [pyg-team/pytorch_geometric#9327] [General]
Hello, I recently tried adding some nodes to a graph to observe changes in the original nodes, such as from A to B, where A is the inserted node. I want to see the changes in B, but I found that using GraphSAGE, the output of node B does not change. I also couldn't find the controls for the sampling operation in GraphSAGE. Could you explain which part controls the node sampling operation in GraphSAGE?

**Comment by wsad1** [comment]
The sampling is controlled by a loader, you can use `NeighborLoader`. See this [example](https://github.com/pyg-team/pytorch_geometric/blob/e9648df16dcb6dde0e09b5736b1b2da5d68db2ad/examples/ogbn_products_sage.py#L20) for more details.

**Reply by 1234238** [reply]
If I don't utilize the NeighborLoader, does SAGEConv inherently lack sampling? Why is it that when I insert an edge from A to B for a node B, altering the features of A appears to have no impact on B? What could be the reason behind this?

**Reply by rusty1s** [reply]
Yes, `SAGEConv` does not sample, it is just the layer implementation. Sampling is performed during data loading. What do you mean by altering the features of `A`? Do you mind sharing a small example to illustrate?

**Reply by 1234238** [reply]
> Yes, `SAGEConv` does not sample, it is just the layer implementation. Sampling is performed during data loading. What do you mean by altering the features of `A`? Do you mind sharing a small example to illustrate?

Thank you for your reply. I apologize for not being clear about my question. For example, I currently have a graph containing three nodes [a, b, c], with connection relationships being [b -> a, c -> a]. Then, the message passing from nodes b and c will affect the features of node a (although this is somewhat different from GCN).

I want to add a new node d and include the connection [d -> a]. At this point, the nodes become [a, b, c, d] and the connection relationships become [b -> a, c -> a, d -> a]. I hope that after adding node d, the features of node a will change after node aggregation (this is a kind of node insertion attack). However, it's strange that in SAGE, sometimes adding node d does not affect node a (but GAT and GCN will have a significant impact), and the classification probability for the node classification task does not change at all. I initially thought that node d was excluded from the sampling nodes, but now it seems that this is not the reason.

**Reply by rusty1s** [reply]
Are features of `b`, `c` and `d` equivalent? This is the only reason why node embeddings in `SAGEConv` would not change.

**Reply by 1234238** [reply]
> Are features of `b`, `c` and `d` equivalent? This is the only reason when node embeddings in `SAGEConv` would not change.

Thank you very much for your reply! I will check my code again.
